drivers/usbc/tcpc/ucpd_stm32.c

```diff
@@ -0,0 +1,1463 @@
+			}
+		}
+		break;
+
+	case STATE_ACTIVE_CRC:
+		if (info->evt & (UCPD_EVT_TX_MSG_SUCCESS |
+				 UCPD_EVT_TX_MSG_FAIL |
+				 UCPD_EVT_TX_MSG_DISC)) {
+			ucpd_set_tx_state(info->dev, STATE_IDLE);
+			if (info->evt & UCPD_EVT_TX_MSG_FAIL) {
+				LOG_INF("ucpd: Failed to send GoodCRC!");
+			} else if (info->evt & UCPD_EVT_TX_MSG_DISC) {
+				LOG_INF("ucpd: GoodCRC message discarded!");
+			}
+		}
+		break;
+
+	case STATE_WAIT_CRC_ACK:
+		if (info->evt & UCPD_EVT_RX_GOOD_CRC &&
+		    data->ucpd_crc_id == data->msg_id_match) {
+			/* GoodCRC with matching ID was received */
+			info->handler(info->dev, info->data,
+				      TCPC_ALERT_TRANSMIT_MSG_SUCCESS);
+			ucpd_set_tx_state(info->dev, STATE_IDLE);
+		} else if (info->evt & UCPD_EVT_RX_GOOD_CRC) {
+			/* GoodCRC w/out match or timeout waiting */
+			if (data->tx_retry_count < data->tx_retry_max) {
+				ucpd_set_tx_state(info->dev, STATE_ACTIVE_TCPM);
+				msg_src = TX_MSG_TCPM;
+				data->tx_retry_count++;
+			} else {
+				ucpd_set_tx_state(info->dev, STATE_IDLE);
+				info->handler(info->dev, info->data,
+					      TCPC_ALERT_TRANSMIT_MSG_FAILED);
+			}
+		} else if (info->evt & UCPD_EVT_RX_MSG) {
+			/*
+			 * In the case of a collsion, it's possible the port
+			 * partner may not send a GoodCRC and instead send the
+			 * message that was colliding. If a message is received
+			 * in this state, then treat it as a discard from an
+			 * incoming message.
+			 */
+			info->handler(info->dev, info->data,
+				      TCPC_ALERT_TRANSMIT_MSG_DISCARDED);
+			ucpd_set_tx_state(info->dev, STATE_IDLE);
+		}
+		break;
+
+	case STATE_HARD_RESET:
+		if (info->evt & UCPD_EVT_HR_DONE) {
+			/* HR complete, reset tx state values */
+			ucpd_set_tx_state(info->dev, STATE_IDLE);
+			data->ucpd_tx_request = 0;
+			data->tx_retry_count = 0;
+		} else if (info->evt & UCPD_EVT_HR_FAIL) {
+			ucpd_set_tx_state(info->dev, STATE_IDLE);
+			data->ucpd_tx_request = 0;
+			data->tx_retry_count = 0;
+		}
+		break;
+	}
+
+	/* If msg_src is valid, then start transmit */
+	if (msg_src > TX_MSG_NONE) {
+		ucpd_start_transmit(info->dev, msg_src);
+	}
+}
+
+/**
+ * @brief Alert handler
+ */
+static void ucpd_alert_handler(struct k_work *item)
+{
+	struct alert_info *info = CONTAINER_OF(item, struct alert_info, work);
+	struct tcpc_data *data = info->dev->data;
+
+	if (info->evt & UCPD_EVT_EVENT_CC) {
+		info->handler(info->dev, info->data, TCPC_ALERT_CC_STATUS);
+	}
+
+	if (info->evt & UCPD_EVT_HARD_RESET_RECEIVED) {
+		info->handler(info->dev, info->data,
+			      TCPC_ALERT_HARD_RESET_RECEIVED);
+	}
+
+	if (info->evt & UCPD_EVT_RX_MSG) {
+		info->handler(info->dev, info->data, TCPC_ALERT_MSG_STATUS);
+	}
+
+	/*
+	 * USB-PD messages are intiated in TCPM stack (PRL
+	 * layer). However, GoodCRC messages are initiated within the
+	 * UCPD driver based on USB-PD rx messages. These 2 types of
+	 * transmit paths are managed via events.
+	 *
+	 * UCPD generated GoodCRC messages, are the priority path as
```

Comment:
Does that mean the stm32 will auto ack the Rx messages with good crc without driver intervention?
if so ... make sense as the goodcrc timing requirement is ~1msec which is hard to achieve with driver code.

Reply:
The STM32 UCPD does not automatically send a GoodCRC when a message is received, so it must be done in the driver. From tests, I've seen the time vary from ~450 to 800uS.  The spec states that max time is 1100uS

Reply:
sending goodcrc from a work item is not a good idea in my opinion.
this work can be blocked for more than 1msec or for a long time if the previous work (by some other module/app)  did not complete its work with in this time.
Note: system work queue will run in sequence order in which the works are queued. Hence if previous work sleep/block for >1msec then the goodcrc which is expected to send within 1msec may not be sent within that time.
This piece of the code might be working now as no other task in your platform blocking work queue more than 1msec.

Reply:
I changed the code so that the GoodCRC is sent from the ISR instead of the work queue